src/connector/src/source/cdc/external/sql_server.rs

```diff
@@ -0,0 +1,480 @@
+    pub change_lsn: String,
+    pub commit_lsn: String,
+}
+
+// only compare the lsn field
+impl PartialOrd for SqlServerOffset {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        match self.change_lsn.partial_cmp(&other.change_lsn) {
+            Some(Ordering::Equal) => self.commit_lsn.partial_cmp(&other.commit_lsn),
+            other => other,
+        }
+    }
+}
+
+impl SqlServerOffset {
+    pub fn parse_debezium_offset(offset: &str) -> ConnectorResult<Self> {
+        let dbz_offset: DebeziumOffset = serde_json::from_str(offset)
+            .with_context(|| format!("invalid upstream offset: {}", offset))?;
+
+        Ok(Self {
+            change_lsn: dbz_offset
+                .source_offset
+                .change_lsn
+                .context("invalid sql server change_lsn")?,
+            commit_lsn: dbz_offset
+                .source_offset
+                .commit_lsn
+                .context("invalid sql server commit_lsn")?,
+        })
+    }
+}
+
+pub struct SqlServerExternalTable {
+    column_descs: Vec<ColumnDesc>,
+    pk_names: Vec<String>,
+}
+
+impl SqlServerExternalTable {
+    pub async fn connect(config: ExternalTableConfig) -> ConnectorResult<Self> {
+        tracing::debug!("connect to sql server");
+
+        let mut client_config = Config::new();
+
+        client_config.host(&config.host);
+        client_config.database(&config.database);
+        client_config.port(config.port.parse::<u16>().unwrap());
+        client_config.authentication(tiberius::AuthMethod::sql_server(
+            &config.username,
+            &config.password,
+        ));
+        client_config.trust_cert();
+
+        let tcp = TcpStream::connect(client_config.get_addr()).await?;
+        tcp.set_nodelay(true)?;
+
+        let mut client: tiberius::Client<Compat<TcpStream>> =
+            match tiberius::Client::connect(client_config, tcp.compat_write()).await {
+                // Connection successful.
+                Ok(client) => Ok(client),
+                // The server wants us to redirect to a different address
+                Err(Error::Routing { host, port }) => {
+                    let mut config = Config::new();
+
+                    config.host(&host);
+                    config.port(port);
+                    config
+                        .authentication(tiberius::AuthMethod::sql_server("sa", "YourPassword123"));
+
+                    let tcp = TcpStream::connect(config.get_addr()).await?;
+                    tcp.set_nodelay(true)?;
+
+                    // we should not have more than one redirect, so we'll short-circuit here.
+                    tiberius::Client::connect(config, tcp.compat_write()).await
+                }
+                Err(e) => Err(e),
+            }?;
+
+        let mut column_descs = vec![];
+        let mut pk_names = vec![];
+        {
+            let sql = Query::new(format!(
+                "SELECT * FROM {} WHERE 1 = 0",
```

Comment:
this query looks weird, could you document the idea?

Reply:
added.